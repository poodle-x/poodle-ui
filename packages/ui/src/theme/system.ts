import {
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	compose,
	createSystem,
	getMode,
	FontsProps,
	ColorsProps,
	SpacingProps,
	LayoutProps,
	BordersProps,
	bordersSystem,
	flexSystem,
	FlexProps,
	pseudosSystem,
	PseudosProps,
	shadowSystem,
	ShadowProps,
	GridProps,
	gridSystem,
	ListProps,
	listSystem,
	SystemConfigList,
	ResponsiveProp,
} from "@poodle/system";
import { CSSSystem } from "../styled";
import { StandardThemeConfig } from "./index";
import * as CSS from "csstype";

const additionalSystem: SystemConfigList = {
	objectFit: true,
	objectPosition: true,
	transition: true,
	cursor: true,
	pointerEvents: true,
};

export interface AdditionalSystemProps {
	/**
	 * The CSS `object-fit` property
	 */
	objectFit?: ResponsiveProp<CSS.Property.ObjectFit>;
	/**
	 * The CSS `object-fit` property
	 */
	objectPosition?: ResponsiveProp<CSS.Property.ObjectPosition>;
	/**
	 * The CSS `transition` property
	 */
	transition?: ResponsiveProp<CSS.Property.Transition>;
	/**
	 * The CSS `cursor` property
	 */
	cursor?: ResponsiveProp<CSS.Property.Cursor>;
	/**
	 * The CSS `cursor` property
	 */
	pointerEvents?: ResponsiveProp<CSS.Property.PointerEvents>;
}

const allSystem = compose(
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	bordersSystem,
	flexSystem,
	pseudosSystem,
	shadowSystem,
	gridSystem,
	listSystem,
	additionalSystem
);

export const system = createSystem(allSystem);

interface CssSystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		GridProps,
		ListProps {}

export interface SystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		PseudosProps<CssSystemProps>,
		GridProps,
		ListProps,
		AdditionalSystemProps {}

const additionalProps = [
	"textStyle",
	"variant",
	"themeExtend",
	"theme",
	"sizeStyle",
	"colorStyle",
];

export function removeSystemProps(props: { [key: string]: any }) {
	let resultProps = {};
	Object.keys(props).forEach((prop) => {
		let found = false;
		for (const name in allSystem) {
			if (prop === name || additionalProps.indexOf(prop) >= 0) {
				found = true;
				break;
			}
		}

		if (!found && ["sx"].indexOf(prop) === -1) {
			resultProps = {
				...resultProps,
				[prop]: props[prop],
			};
		}
	});

	return resultProps;
}

export function createCSSSystemStandard<Key extends string>(data: {
	key: Key;
	config?: StandardThemeConfig<Key, any>;
	props: { [key: string]: any };
	base?: CSSSystem["base"];
}): CSSSystem {
	const { key, props, config, base } = data;

	const { variant = "", sizeStyle = "", colorStyle = "" } = props;

	const mode = getMode(props);

	const overrideClassName = config?.overrideClasses?.[key];

	return {
		base,
		overrideClassName,
		applies: [
			config?.styles?.[key],
			config?.modes?.[mode]?.styles?.[key],
			config?.variants?.[variant]?.[key],
			config?.modes?.[mode]?.variants?.[variant]?.[key],
			config?.sizes?.[sizeStyle]?.[key],
			config?.modes?.[mode]?.sizes?.[sizeStyle]?.[key],
			config?.colorStyles?.[colorStyle]?.[key],
			config?.modes?.[mode]?.colorStyles?.[colorStyle]?.[key],
		],
		override: config?.overrides?.[key],
	};
}

export * from "@poodle/system";
