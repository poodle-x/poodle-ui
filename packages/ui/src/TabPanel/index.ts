import { TabPanel, TabPanelProps } from "./TabPanel";
import { StandardThemeConfig } from "../theme";

export * from "./TabPanel";

export default TabPanel;

export type TabPanelStyleKeys = "root";

export interface TabPanelThemeConfig
	extends StandardThemeConfig<TabPanelStyleKeys, TabPanelProps> {}
