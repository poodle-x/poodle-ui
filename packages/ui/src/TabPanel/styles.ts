import { createCSSSystemStandard } from "../theme";
import { CSSSystem } from "../styled";
import { TabPanelProps } from "./TabPanel";

export function Root(props: TabPanelProps): CSSSystem {
	const { theme } = props;

	return createCSSSystemStandard({
		key: "root",
		config: theme?.TabPanel,
		props,
		base: [
			{
				paddingY: "scale-4",
				paddingX: "scale-4",
				outline: "none",
			},
		],
	});
}
