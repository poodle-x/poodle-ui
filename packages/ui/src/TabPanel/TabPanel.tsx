import React from "react";
import Box, { BoxProps } from "../Box";
import { StandardComponentProps, ThemeConfig } from "../theme";
import useDefaultProps from "../utils/useDefaultProps";
import uSetRef from "../utils/setRef";
import { useClassNames } from "../styled";
import * as styles from "./styles";
import { TabValue } from "../hooks/useTabsState";
import useTabPanel from "../hooks/useTabPanel";

export interface LocalTabPanelProps {
	value: TabValue;
	/**
	 * Does not unmount the children when the panel is not selected.
	 */
	keepMountChildren?: boolean;
	children?: React.ReactNode;
}

export interface TabPanelProps
	extends Omit<BoxProps, "value">,
		StandardComponentProps,
		LocalTabPanelProps {}

function getDefaultProps(theme?: ThemeConfig) {
	return theme?.TabPanel?.defaultProps;
}

export const TabPanel: React.ForwardRefExoticComponent<
	React.PropsWithoutRef<TabPanelProps> & React.RefAttributes<HTMLElement>
> = React.forwardRef<HTMLElement, TabPanelProps>((_props, ref) => {
	const props = useDefaultProps<TabPanelProps>(_props, {
		themeDefaultProps: getDefaultProps,
	});

	const {
		children,
		className,
		id,
		value,
		keepMountChildren,
		...otherProps
	} = props;

	const { props: htmlProps, selected, setRef } = useTabPanel({ id, value });

	const classes = useClassNames({
		props,
		lists: {
			root: {
				classNames: ["poodle-tabpanel", styles.Root, className],
			},
		},
	});

	const composeRef = React.useCallback(
		(node: HTMLElement) => {
			setRef(node);
			uSetRef(ref, node);
		},
		[setRef, ref]
	);

	return (
		<Box
			{...htmlProps}
			{...otherProps}
			className={classes.root}
			ref={composeRef}
		>
			{selected || keepMountChildren ? children : null}
		</Box>
	);
});

TabPanel.displayName = "PoodleTabPanel";
