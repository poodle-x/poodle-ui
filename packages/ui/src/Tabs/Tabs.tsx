import React from "react";
import { StandardComponentProps, ThemeConfig } from "../theme";
import useDefaultProps from "../utils/useDefaultProps";
import { useClassNames } from "../styled";
import * as styles from "./styles";
import useTabsState, {
	TabOrientation,
	TabsContext,
	TabsContextValue,
	TabValue,
} from "../hooks/useTabsState";
import Box, { BoxProps } from "../Box";

export { TabOrientation, TabsContext, TabsContextValue, TabValue };

export interface LocalTabsProps {
	/**
	 * Send value to control active tab. By default <Tabs /> component
	 * use internal state to control active tab.
	 */
	value?: TabValue;
	/**
	 * Change tabs horizontal orientation or vertical orientation.
	 * Default is horizontal orientation.
	 */
	orientation?: TabOrientation;
	onChangeValue?: (
		value: TabValue,
		event: React.SyntheticEvent | KeyboardEvent
	) => void;
	children?: React.ReactNode;
}

export interface TabsProps
	extends Omit<BoxProps, "value">,
		StandardComponentProps,
		LocalTabsProps {}

function getDefaultProps(theme?: ThemeConfig) {
	return theme?.Tabs?.defaultProps;
}

export const Tabs: React.ForwardRefExoticComponent<
	React.PropsWithoutRef<TabsProps> & React.RefAttributes<HTMLElement>
> = React.forwardRef<HTMLElement, TabsProps>((_props, ref) => {
	const props = useDefaultProps<TabsProps>(_props, {
		themeDefaultProps: getDefaultProps,
	});

	const {
		children,
		className,
		value,
		onChangeValue,
		orientation,
		...otherProps
	} = props;

	const {
		setUnControlledValue,
		isUncontrolled,
		value: calcValue,
		handleUnregister,
		handleRegister,
		listTabPanel,
		listTab,
	} = useTabsState({ propValue: value });

	const classes = useClassNames({
		props,
		lists: React.useMemo(() => {
			return {
				root: {
					classNames: ["poodle-tabs", styles.Root, className],
				},
			};
		}, [className]),
	});

	const handleChangeValue: TabsContextValue["onChange"] = React.useCallback(
		(v, e) => {
			onChangeValue && onChangeValue(v, e);
			if (isUncontrolled) {
				setUnControlledValue(v);
			}
		},
		[onChangeValue, isUncontrolled, setUnControlledValue]
	);

	return (
		<TabsContext.Provider
			value={{
				listTab,
				listTabPanel,
				value: calcValue,
				register: handleRegister,
				unregister: handleUnregister,
				onChange: handleChangeValue,
				orientation: orientation || TabOrientation.HORIZONTAL,
			}}
		>
			<Box {...otherProps} className={classes.root} ref={ref}>
				{children}
			</Box>
		</TabsContext.Provider>
	);
});

Tabs.displayName = "PoodleTabs";
