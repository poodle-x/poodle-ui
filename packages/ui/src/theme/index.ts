import React from "react";
import {
	ThemeValue,
	SystemThemeConfig,
	BreakpointsThemeConfig,
	SpacingThemeConfig,
	ColorsThemeConfig,
	FontsThemeConfig,
	BordersThemeConfig,
	ShadowThemeConfig,
	LayoutsThemeConfig,
} from "@poodle/system";

import { CSSObjectSystem, Stylesheet } from "../styled";
import { BoxThemeConfig } from "../Box";
import { ButtonThemeConfig } from "../Button";
import { ColumnsThemeConfig } from "../Columns";
import { ColumnThemeConfig } from "../Column";
import { ContainerThemeConfig } from "../Container";
import { TableThemeConfig } from "../Table";
import { TableHeadThemeConfig } from "../TableHead";
import { TableBodyThemeConfig } from "../TableBody";
import { TableRowThemeConfig } from "../TableRow";
import { TableCellThemeConfig } from "../TableCell";
import { VisuallyHiddenThemeConfig } from "../VisuallyHidden";
import { CheckControlThemeConfig } from "../CheckControl";
import { CheckboxThemeConfig } from "../Checkbox";
import { RadioThemeConfig } from "../Radio";
import { InputThemeConfig } from "../Input";
import { InputAdornmentThemeConfig } from "../InputAdornment";
import { IconButtonThemeConfig } from "../IconButton";
import { ModalThemeConfig } from "../Modal";
import { ModalHeaderThemeConfig } from "../ModalHeader";
import { ModalBodyThemeConfig } from "../ModalBody";
import { ModalFooterThemeConfig } from "../ModalFooter";
import { SelectThemeConfig } from "../Select";
import { TabsThemeConfig } from "../Tabs";
import { TabThemeConfig } from "../Tab";
import { TabListThemeConfig } from "../TabList";
import { TabPanelThemeConfig } from "../TabPanel";

export { Stylesheet };

export { useTheme } from "./useTheme";

export type DirectionThemeConfig = "rtl" | "ltr";

export interface GlobalThemeConfig {
	styles?: ThemeValue<CSSObjectSystem>;
}

export type IconType = React.FC<
	React.PropsWithoutRef<React.SVGProps<SVGSVGElement>> &
		React.RefAttributes<SVGSVGElement>
>;

export interface BaseIcons {
	RadioButton?: IconType;
	RadioCheckedButton?: IconType;
	CheckboxButton?: IconType;
	CheckboxCheckedButton?: IconType;
	CheckboxIndeterminateButton?: IconType;
	ArrowDown?: IconType;
}

export interface ThemeConfig extends SystemThemeConfig {
	global?: GlobalThemeConfig;
	direction?: ThemeValue<DirectionThemeConfig>;
	baseIcons?: BaseIcons;
	Box?: BoxThemeConfig;
	Button?: ButtonThemeConfig;
	Columns?: ColumnsThemeConfig;
	Column?: ColumnThemeConfig;
	Container?: ContainerThemeConfig;
	Table?: TableThemeConfig;
	TableHead?: TableHeadThemeConfig;
	TableBody?: TableBodyThemeConfig;
	TableRow?: TableRowThemeConfig;
	TableCell?: TableCellThemeConfig;
	VisuallyHidden?: VisuallyHiddenThemeConfig;
	CheckControl?: CheckControlThemeConfig;
	Checkbox?: CheckboxThemeConfig;
	Radio?: RadioThemeConfig;
	Input?: InputThemeConfig;
	InputAdornment?: InputAdornmentThemeConfig;
	IconButton?: IconButtonThemeConfig;
	Modal?: ModalThemeConfig;
	ModalHeader?: ModalHeaderThemeConfig;
	ModalBody?: ModalBodyThemeConfig;
	ModalFooter?: ModalFooterThemeConfig;
	Select?: SelectThemeConfig;
	Tabs?: TabsThemeConfig;
	Tab?: TabThemeConfig;
	TabList?: TabListThemeConfig;
	TabPanel?: TabPanelThemeConfig;
}

export const defaultGlobal = (
	overrideTheme: ThemeConfig
): GlobalThemeConfig => {
	return {
		styles: overrideTheme.global?.styles,
	};
};

export const defaultBreakpoints = (
	overrideTheme: ThemeConfig
): BreakpointsThemeConfig => {
	return {
		scale: ["sm", "md", "lg", "xl"],
		...overrideTheme.breakpoints,
		sets: {
			sm: 576,
			md: 768,
			lg: 992,
			xl: 1200,
			xxl: 1400,
			...overrideTheme.breakpoints?.sets,
		},
	};
};

export const defaultSpacing = (
	overrideTheme: ThemeConfig
): SpacingThemeConfig => {
	return {
		scale: 4,
		...overrideTheme.spacing,
		sets: {
			s: 4,
			m: 8,
			l: 16,
			...overrideTheme.spacing?.sets,
		},
	};
};

export const defaultFonts = (overrideTheme: ThemeConfig): FontsThemeConfig => {
	return {
		scale: "1rem",
		...overrideTheme.fonts,
		base: {
			fontFamily:
				'-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
			fontStyle: "normal",
			fontWeight: "normal",
			lineHeight: "125%",
			fontSize: "1rem",
			...overrideTheme.fonts?.base,
		},
		weights: {
			normal: 400,
			semibold: 600,
			bold: 700,
			...overrideTheme.fonts?.weights,
		},
		sets: {
			input: {
				fontSize: "1rem",
				fontWeight: "normal",
				lineHeight: "125%",
			},
			button: {
				fontSize: "1rem",
				fontWeight: "bold",
				lineHeight: "150%",
			},
			label: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			helper: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelMedium: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1rem",
				lineHeight: "110%",
			},
			checkControlLabelSmall: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelLarge: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1.125rem",
				lineHeight: "110%",
			},
			modalTitle: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "1.5rem",
				lineHeight: "133.33%",
			},
			...overrideTheme.fonts?.sets,
		},
	};
};

export const defaultColor = (overrideTheme: ThemeConfig): ColorsThemeConfig => {
	return {
		modes: {
			...overrideTheme.colors?.modes,
		},
		sets: {
			text: "#282828",
			disabledText: "#A6A6A6",

			primary: {
				light: "#D1EBFF",
				base: "#00478A",
				dark: "#003A70",
				textOnLight: "#00478A",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			negative: {
				base: "#AF1308",
				dark: "#8D1007",
				light: "#FCF3F3",
				textOnLight: "#AF1308",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			positive: {
				base: "#006122",
				dark: "#004D1B",
				light: "#E5FFEE",
				textOnLight: "#006122",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			warn: {
				base: "#FFBB33",
				dark: "#FAAB00",
				light: "#FFF1DB",
				textOnLight: "#282828",
				textOnBase: "#282828",
				textOnDark: "#282828",
			},

			bg: {
				base: "#F8F8F8",
				100: "#E0E0E0",
				textOn100: "#282828",
				200: "#CCCCCC",
				textOn200: "#282828",
			},

			modal: {
				base: "#F8F8F8",
				footer: "#F0F0F0",
			},

			border: {
				base: "#CCCCCC",
			},

			...overrideTheme.colors?.sets,
		},
	};
};

export const defaultBorder = (
	overrideTheme: ThemeConfig
): BordersThemeConfig => {
	return {
		sets: {
			base: {
				style: "solid",
				width: "1px",
				color: "border",
			},
			...overrideTheme?.borders?.sets,
		},
		radius: {
			base: "4px",
			...overrideTheme?.borders?.radius,
		},
	};
};

export const defaultBoxShadow = (
	overrideTheme: ThemeConfig
): ShadowThemeConfig => {
	return {
		boxShadowSets: {
			lv1: "1px 2px 2px -1px rgba(0,0,0,0.4)",
			lv2: "2px 2px 5px 0px rgba(0,0,0,0.4)",
			lv3: "4px 4px 5px 0px rgba(0,0,0,0.4)",
			...overrideTheme?.shadows?.boxShadowSets,
		},
	};
};

export const defaultLayouts = (
	overrideTheme: ThemeConfig
): LayoutsThemeConfig => {
	return {
		zIndices: {
			modal: 2000,
			...overrideTheme?.layouts?.zIndices,
		},
	};
};

export function createTheme(overrideTheme: ThemeConfig = {}): ThemeConfig {
	return {
		direction: "ltr",
		...overrideTheme,
		global: defaultGlobal(overrideTheme),
		spacing: defaultSpacing(overrideTheme),
		breakpoints: defaultBreakpoints(overrideTheme),
		colors: defaultColor(overrideTheme),
		fonts: defaultFonts(overrideTheme),
		borders: defaultBorder(overrideTheme),
		shadows: defaultBoxShadow(overrideTheme),
		layouts: defaultLayouts(overrideTheme),
	};
}

export type StylesConfig<T extends string> = {
	[K in T]: ThemeValue<CSSObjectSystem>;
};

export interface StandardComponentProps {
	variant?: string;
	themeExtend?: ThemeConfig;
	theme?: ThemeConfig;
	sizeStyle?: string;
	colorStyle?: string;
}

export interface StandardThemeConfig<
	T extends string,
	Props extends { [key: string]: any }
> {
	defaultProps?: Partial<Props>;
	styles?: Partial<StylesConfig<T>>;
	variants?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	sizes?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	colorStyles?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	modes?: {
		[key: string]: {
			styles?: Partial<StylesConfig<T>>;
			variants?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			sizes?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			colorStyles?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
		};
	};
	overrides?: Partial<StylesConfig<T>>;
	overrideClasses?: Partial<
		{
			[K in T]: boolean | string;
		}
	>;
}

export * from "./system";
