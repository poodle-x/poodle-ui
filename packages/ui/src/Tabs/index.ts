import { Tabs, TabsProps } from "./Tabs";
import { StandardThemeConfig } from "../theme";
import { TabValue, TabsContext, TabsContextValue } from "../hooks/useTabsState";

export * from "./Tabs";

export default Tabs;

export type TabsStyleKeys = "root" | "list";

export interface TabsThemeConfig
	extends StandardThemeConfig<TabsStyleKeys, TabsProps> {}

export { TabValue, TabsContext, TabsContextValue };
