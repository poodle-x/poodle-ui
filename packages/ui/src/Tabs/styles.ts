import { createCSSSystemStandard } from "../theme";
import { CSSSystem } from "../styled";
import { TabsProps } from "./Tabs";
import { TabsStyleKeys } from "./index";
import { TabOrientation } from "../hooks/useTabsState";

export function Root(props: TabsProps): CSSSystem {
	const { theme, orientation = TabOrientation.HORIZONTAL } = props;

	return createCSSSystemStandard<TabsStyleKeys>({
		key: "root",
		config: theme?.Tabs,
		props,
		base: [
			{
				display: "flex",
				flexDirection:
					orientation === TabOrientation.VERTICAL ? "row" : "column",
			},
		],
	});
}
