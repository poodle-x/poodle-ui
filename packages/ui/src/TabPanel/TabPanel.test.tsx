import React from "react";
import { render } from "../utils/test";
import TabPanel from "./";
import { ThemeConfig } from "../theme";
import { Tabs } from "../Tabs";

describe("<TabPanel />", () => {
	it("should render TabPanel correctly", async () => {
		const { findByRole, rerender } = render(
			<Tabs>
				<TabPanel value="text">TabPanel</TabPanel>
			</Tabs>
		);
		expect(await findByRole("tabpanel", { hidden: true })).toMatchSnapshot();

		rerender(
			<Tabs value="text">
				<TabPanel value="text">TabPanel</TabPanel>
			</Tabs>
		);

		expect(await findByRole("tabpanel")).toMatchSnapshot();
	});

	it("ref should work", () => {
		const ref = React.createRef<HTMLDivElement>();

		render(
			<Tabs>
				<TabPanel ref={ref} value="text">
					TabPanel
				</TabPanel>
			</Tabs>
		);
		expect(ref.current).toMatchSnapshot();
	});

	it("should render Box props correctly", () => {
		const { container } = render(
			<Tabs>
				<TabPanel value="text" id="id" mt="20px">
					TabPanel
				</TabPanel>
			</Tabs>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	describe("theming", () => {
		const themeTest: ThemeConfig = {
			mode: "mode",
			TabPanel: {
				styles: {
					root: {
						mt: 10,
						letterSpacing: 250,
					},
				},
				variants: {
					custom: {
						root: {
							top: 0,
							textIndent: 1,
						},
					},
				},
			},
		};

		it("should extend styles in global", () => {
			const { container } = render(
				<Tabs>
					<TabPanel value="text" variant="custom">
						TabPanel
					</TabPanel>
				</Tabs>,
				{
					theme: themeTest,
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should extend styles in local", () => {
			const { container } = render(
				<Tabs>
					<TabPanel value="text" variant="custom" themeExtend={themeTest}>
						TabPanel
					</TabPanel>
				</Tabs>
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should render with default props correctly", () => {
			const { container } = render(
				<Tabs>
					<TabPanel value="text">TabPanel</TabPanel>
				</Tabs>,
				{
					theme: {
						TabPanel: {
							defaultProps: {
								id: "id",
								className: "hello",
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should disable classes", () => {
			const { container } = render(
				<Tabs>
					<TabPanel value="text">TabPanel</TabPanel>
				</Tabs>,
				{
					theme: {
						TabPanel: {
							overrideClasses: {
								root: true,
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override classes", () => {
			const { container } = render(
				<Tabs>
					<TabPanel value="text">TabPanel</TabPanel>
				</Tabs>,
				{
					theme: {
						TabPanel: {
							overrideClasses: {
								root: "morning",
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override styles", () => {
			const { container } = render(
				<Tabs>
					<TabPanel value="text">TabPanel</TabPanel>
				</Tabs>,
				{
					theme: {
						TabPanel: {
							overrides: {
								root: {
									mt: 50,
								},
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
