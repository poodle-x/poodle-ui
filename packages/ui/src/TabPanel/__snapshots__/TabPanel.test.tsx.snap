// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabPanel /> ref should work 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 16px;
  padding-right: 16px;
  outline: none;
}

<div
  class="poodle-box c0 poodle-tabpanel c1 c2"
  hidden=""
  id="poodle-tab-panel-1"
  role="tabpanel"
  tabindex="0"
/>
`;

exports[`<TabPanel /> should render Box props correctly 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 16px;
  padding-right: 16px;
  outline: none;
}

.c5 {
  margin-top: 20px;
}

<div
  class="poodle-box c0 poodle-tabs c1 c2"
>
  <div
    class="poodle-box c0 poodle-tabpanel c4 c5"
    hidden=""
    id="id"
    role="tabpanel"
    tabindex="0"
  />
</div>
`;

exports[`<TabPanel /> should render TabPanel correctly 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 16px;
  padding-right: 16px;
  outline: none;
}

<div
  class="poodle-box c0 poodle-tabpanel c1 c2"
  hidden=""
  id="poodle-tab-panel-1"
  role="tabpanel"
  tabindex="0"
/>
`;

exports[`<TabPanel /> should render TabPanel correctly 2`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 16px;
  padding-right: 16px;
  outline: none;
}

<div
  class="poodle-box c0 poodle-tabpanel c1 c2"
  id="poodle-tab-panel-1"
  role="tabpanel"
  tabindex="0"
>
  TabPanel
</div>
`;

exports[`<TabPanel /> theming should disable classes 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  class="poodle-box c0 poodle-tabs c1 c2"
>
  <div
    class="poodle-box c0 poodle-tabpanel c2"
    hidden=""
    id="poodle-tab-panel-1"
    role="tabpanel"
    tabindex="0"
  />
</div>
`;

exports[`<TabPanel /> theming should extend styles in global 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 16px;
  padding-right: 16px;
  outline: none;
  margin-top: 10px;
  letter-spacing: 250px;
  top: 0;
  text-indent: 1px;
}

<div
  class="poodle-box c0 poodle-tabs c1 c2"
>
  <div
    class="poodle-box c0 poodle-tabpanel c4 c2"
    hidden=""
    id="poodle-tab-panel-1"
    role="tabpanel"
    tabindex="0"
  />
</div>
`;

exports[`<TabPanel /> theming should extend styles in local 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 16px;
  padding-right: 16px;
  outline: none;
  margin-top: 10px;
  letter-spacing: 250px;
  top: 0;
  text-indent: 1px;
}

<div
  class="poodle-box c0 poodle-tabs c1 c2"
>
  <div
    class="poodle-box c0 poodle-tabpanel c4 c2"
    hidden=""
    id="poodle-tab-panel-1"
    role="tabpanel"
    tabindex="0"
  />
</div>
`;

exports[`<TabPanel /> theming should override classes 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  class="poodle-box c0 poodle-tabs c1 c2"
>
  <div
    class="poodle-box c0 poodle-tabpanel morning c2"
    hidden=""
    id="poodle-tab-panel-1"
    role="tabpanel"
    tabindex="0"
  />
</div>
`;

exports[`<TabPanel /> theming should override styles 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  margin-top: 50px;
}

<div
  class="poodle-box c0 poodle-tabs c1 c2"
>
  <div
    class="poodle-box c0 poodle-tabpanel c4 c2"
    hidden=""
    id="poodle-tab-panel-1"
    role="tabpanel"
    tabindex="0"
  />
</div>
`;

exports[`<TabPanel /> theming should render with default props correctly 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 16px;
  padding-right: 16px;
  outline: none;
}

<div
  class="poodle-box c0 poodle-tabs c1 c2"
>
  <div
    class="poodle-box c0 poodle-tabpanel c4 hello c2"
    hidden=""
    id="id"
    role="tabpanel"
    tabindex="0"
  />
</div>
`;
